## The Pandas Series
A series is similar to a 1-D numpy array, and contains scalar values of the same type (numeric, character, datetime etc.). A dataframe is simply a table where each column is a pandas series.

Creating Pandas Series
Series are one-dimensional array-like structures, though unlike numpy arrays, they often contain non-numeric data (characters, dates, time, booleans etc.)

You can create pandas series from array-like objects using pd.Series().##

In [4]:
# import pandas, pd is an alias
import pandas as pd
import numpy as np

# Creating a numeric pandas series
s = pd.Series([2, 4, 5, 6, 9])
print(s)
print(type(s))

0    2
1    4
2    5
3    6
4    9
dtype: int64
<class 'pandas.core.series.Series'>


In [2]:
# Note that each element in the Series has an index, and the index starts at 0 as usual.

# creating a series of characters
# notice that the 'dtype' here is 'object'
char_series = pd.Series(['a', 'b', 'af'])
char_series

0     a
1     b
2    af
dtype: object

In [4]:
# creating a series of type datetime
date_series = pd.date_range(start = '11-09-2023', end = '12-12-2023')
date_series

DatetimeIndex(['2023-11-09', '2023-11-10', '2023-11-11', '2023-11-12',
               '2023-11-13', '2023-11-14', '2023-11-15', '2023-11-16',
               '2023-11-17', '2023-11-18', '2023-11-19', '2023-11-20',
               '2023-11-21', '2023-11-22', '2023-11-23', '2023-11-24',
               '2023-11-25', '2023-11-26', '2023-11-27', '2023-11-28',
               '2023-11-29', '2023-11-30', '2023-12-01', '2023-12-02',
               '2023-12-03', '2023-12-04', '2023-12-05', '2023-12-06',
               '2023-12-07', '2023-12-08', '2023-12-09', '2023-12-10',
               '2023-12-11', '2023-12-12'],
              dtype='datetime64[ns]', freq='D')

In [5]:
type(date_series)

pandas.core.indexes.datetimes.DatetimeIndex

In [8]:
# Indexing Series
# Indexing series is exactly same as 1-D numpy arrays - index starts at 0.

s = pd.Series([1,2,3,4])
s

0    1
1    2
2    3
3    4
dtype: int64

In [9]:
s[2:]

2    3
3    4
dtype: int64

In [10]:
# accessing the second and the fourth elements
# note that s[1, 3] will not work, you need to pass the indices [1, 3] as a list inside the original []
s[[1, 3]]

1    2
3    4
dtype: int64

## Explicitly specifying indices
You might have noticed that while creating a series, Pandas automatically indexes it from 0 to (n-1), n being the number of rows. But if we want, we can also explicitly set the index ourselves, using the ‘index’ argument while creating the series using pd.Series()

In [11]:
# Indexing explicitly
pd.Series([0, 1, 2], index = ['a', 'b', 'c'])

a    0
b    1
c    2
dtype: int64

In [12]:
pd.Series([1,2,3],index=pd.date_range('22-11-2018','24-11-2018'))

2018-11-22    1
2018-11-23    2
2018-11-24    3
Freq: D, dtype: int64

In [14]:
# You can also give the index as a sequence or use functions to specify the index
# But always make sure that the number of elements in the index list is equal to the number of elements specified in the series

pd.Series(np.array(range(1,12))**2, index = range(1,12))

1       1
2       4
3       9
4      16
5      25
6      36
7      49
8      64
9      81
10    100
11    121
dtype: int64

## The Pandas Dataframe
Dataframe is the most widely used data-structure in data analysis. It is a table with rows and columns, with rows having an index and columns having meaningful names.

Creating dataframes from dictionaries
There are various ways of creating dataframes, such as creating them from dictionaries, JSON objects, reading from txt, CSV files, etc.

In [19]:
# key become column names
df = pd.DataFrame({
    'FName':['Alade','Aribisala','Dristol','Frank'],
    'LName':['Idris','Ayodeji','Tolani','Salim'],
    'age':[23,24,16,26],
    'occupation':['engineer','doctor','data analyst','Network administrator']

})
df



FName    LName  age             occupation
0      Alade    Idris   23               engineer
1  Aribisala  Ayodeji   24                 doctor
2    Dristol   Tolani   16           data analyst
3      Frank    Salim   26  Network administrator

## Importing CSV data files as pandas dataframes
For the upcoming exercises, we will use a dataset of a retail store having details about the orders placed, customers, product details, sales, profits etc.

In [11]:
# reading a CSV file as a dataframe
market_df = pd.read_csv("market_fact.csv")
market_df

Ord_id  Prod_id   Ship_id    Cust_id      Sales  Discount  \
0     Ord_5446  Prod_16  SHP_7609  Cust_1818   136.8100      0.01   
1     Ord_5406  Prod_13  SHP_7549  Cust_1818    42.2700      0.01   
2     Ord_5446   Prod_4  SHP_7610  Cust_1818  4701.6900      0.00   
3     Ord_5456   Prod_6  SHP_7625  Cust_1818  2337.8900      0.09   
4     Ord_5485  Prod_17  SHP_7664  Cust_1818  4233.1500      0.08   
...        ...      ...       ...        ...        ...       ...   
8394  Ord_5353   Prod_4  SHP_7479  Cust_1798  2841.4395      0.08   
8395  Ord_5411   Prod_6  SHP_7555  Cust_1798   127.1600      0.10   
8396  Ord_5388   Prod_6  SHP_7524  Cust_1798   243.0500      0.02   
8397  Ord_5348  Prod_15  SHP_7469  Cust_1798  3872.8700      0.03   
8398  Ord_5459   Prod_6  SHP_7628  Cust_1798   603.6900      0.00   

      Order_Quantity   Profit  Shipping_Cost  Product_Base_Margin  
0                 23   -30.51           3.60                 0.56  
1                 13     4.56           0.93                 0.54  
2                 26  1148.90           2.50                 0.59  
3                 43   729.34          14.30                 0.37  
4                 35  1219.87          26.30                 0.38  
...              ...      ...            ...                  ...  
8394              28   374.63           7.69                 0.59  
8395              20   -74.03           6.92                 0.37  
8396              39   -70.85           5.35                 0.40  
8397              23   565.34          30.00                 0.62  
8398              47   131.39           4.86                 0.38  

[8399 rows x 10 columns]

Usually, dataframes are imported as CSV files, but sometimes it is more convenient to convert dictionaries into dataframes. For e.g. when the raw data is in a JSON format (which is not uncommon), you can easily convert it into a dictionary, and then into a dataframe.

You will learn how to convert JSON objects to dataframes later.

### Reading and Summarising Dataframes
After you import a dataframe, you'd want to quickly understand its structure, shape, meanings of rows and columns etc. Further, you may want to look at summary statistics - such as mean, percentiles etc.

In [20]:
# Looking at the top and bottom entries of dataframes
market_df.head()

Ord_id  Prod_id   Ship_id    Cust_id    Sales  Discount  Order_Quantity  \
0  Ord_5446  Prod_16  SHP_7609  Cust_1818   136.81      0.01              23   
1  Ord_5406  Prod_13  SHP_7549  Cust_1818    42.27      0.01              13   
2  Ord_5446   Prod_4  SHP_7610  Cust_1818  4701.69      0.00              26   
3  Ord_5456   Prod_6  SHP_7625  Cust_1818  2337.89      0.09              43   
4  Ord_5485  Prod_17  SHP_7664  Cust_1818  4233.15      0.08              35   

    Profit  Shipping_Cost  Product_Base_Margin  
0   -30.51           3.60                 0.56  
1     4.56           0.93                 0.54  
2  1148.90           2.50                 0.59  
3   729.34          14.30                 0.37  
4  1219.87          26.30                 0.38

In [12]:
market_df.tail(10)

Ord_id  Prod_id   Ship_id    Cust_id      Sales  Discount  \
8389  Ord_4444  Prod_13  SHP_6192  Cust_1519   159.4100      0.00   
8390  Ord_5435  Prod_16  SHP_7594  Cust_1798   316.9900      0.04   
8391  Ord_5435   Prod_4  SHP_7594  Cust_1798  1991.8985      0.07   
8392  Ord_5384   Prod_9  SHP_7519  Cust_1798   181.5000      0.08   
8393  Ord_5348   Prod_8  SHP_7470  Cust_1798   356.7200      0.07   
8394  Ord_5353   Prod_4  SHP_7479  Cust_1798  2841.4395      0.08   
8395  Ord_5411   Prod_6  SHP_7555  Cust_1798   127.1600      0.10   
8396  Ord_5388   Prod_6  SHP_7524  Cust_1798   243.0500      0.02   
8397  Ord_5348  Prod_15  SHP_7469  Cust_1798  3872.8700      0.03   
8398  Ord_5459   Prod_6  SHP_7628  Cust_1798   603.6900      0.00   

      Order_Quantity  Profit  Shipping_Cost  Product_Base_Margin  
8389              44   34.68           0.98                 0.52  
8390              47 -276.54           8.37                 0.58  
8391              20   88.36           7.69                 0.58  
8392              43   -6.24           2.50                 0.37  
8393               9   12.61           1.99                 0.44  
8394              28  374.63           7.69                 0.59  
8395              20  -74.03           6.92                 0.37  
8396              39  -70.85           5.35                 0.40  
8397              23  565.34          30.00                 0.62  
8398              47  131.39           4.86                 0.38

Here, each row represents an order placed at a retail store. Notice the index associated with each row - starts at 0 and ends at 8398, implying that there were 8399 orders placed.

In [22]:
# Looking at the datatypes of each column
market_df.info()

# Note that each column is basically a pandas Series of length 8399
# The ID columns are 'objects', i.e. they are being read as characters
# The rest are numeric (floats or int)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8399 entries, 0 to 8398
Data columns (total 10 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Ord_id               8399 non-null   object 
 1   Prod_id              8399 non-null   object 
 2   Ship_id              8399 non-null   object 
 3   Cust_id              8399 non-null   object 
 4   Sales                8399 non-null   float64
 5   Discount             8399 non-null   float64
 6   Order_Quantity       8399 non-null   int64  
 7   Profit               8399 non-null   float64
 8   Shipping_Cost        8399 non-null   float64
 9   Product_Base_Margin  8336 non-null   float64
dtypes: float64(5), int64(1), object(4)
memory usage: 656.3+ KB


In [23]:
# Describe gives you a summary of all the numeric columns in the dataset
market_df.describe()

Sales     Discount  Order_Quantity        Profit  Shipping_Cost  \
count   8399.000000  8399.000000     8399.000000   8399.000000    8399.000000   
mean    1775.878179     0.049671       25.571735    181.184424      12.838557   
std     3585.050525     0.031823       14.481071   1196.653371      17.264052   
min        2.240000     0.000000        1.000000 -14140.700000       0.490000   
25%      143.195000     0.020000       13.000000    -83.315000       3.300000   
50%      449.420000     0.050000       26.000000     -1.500000       6.070000   
75%     1709.320000     0.080000       38.000000    162.750000      13.990000   
max    89061.050000     0.250000       50.000000  27220.690000     164.730000   

       Product_Base_Margin  
count          8336.000000  
mean              0.512513  
std               0.135589  
min               0.350000  
25%               0.380000  
50%               0.520000  
75%               0.590000  
max               0.850000

In [24]:
# column names
market_df.columns

Index(['Ord_id', 'Prod_id', 'Ship_id', 'Cust_id', 'Sales', 'Discount',
       'Order_Quantity', 'Profit', 'Shipping_Cost', 'Product_Base_Margin'],
      dtype='object')

In [25]:
# The number of rows and columns
market_df.shape

(8399, 10)

In [26]:
# we can extract the values of a dataframe as a numpy array using df.values 
market_df.values

array([['Ord_5446', 'Prod_16', 'SHP_7609', ..., -30.51, 3.6, 0.56],
       ['Ord_5406', 'Prod_13', 'SHP_7549', ..., 4.56, 0.93, 0.54],
       ['Ord_5446', 'Prod_4', 'SHP_7610', ..., 1148.9, 2.5, 0.59],
       ...,
       ['Ord_5388', 'Prod_6', 'SHP_7524', ..., -70.85, 5.35, 0.4],
       ['Ord_5348', 'Prod_15', 'SHP_7469', ..., 565.34, 30.0, 0.62],
       ['Ord_5459', 'Prod_6', 'SHP_7628', ..., 131.39, 4.86, 0.38]],
      dtype=object)

## Indices
An important concept in pandas dataframes is that of row indices. By default, each row is assigned indices starting from 0, and are represented at the left side of the dataframe.

In [13]:
market_df.head(20)

Ord_id  Prod_id   Ship_id    Cust_id       Sales  Discount  \
0   Ord_5446  Prod_16  SHP_7609  Cust_1818    136.8100      0.01   
1   Ord_5406  Prod_13  SHP_7549  Cust_1818     42.2700      0.01   
2   Ord_5446   Prod_4  SHP_7610  Cust_1818   4701.6900      0.00   
3   Ord_5456   Prod_6  SHP_7625  Cust_1818   2337.8900      0.09   
4   Ord_5485  Prod_17  SHP_7664  Cust_1818   4233.1500      0.08   
5   Ord_5446   Prod_6  SHP_7608  Cust_1818    164.0200      0.03   
6     Ord_31  Prod_12    SHP_41    Cust_26     14.7600      0.01   
7   Ord_4725   Prod_4  SHP_6593  Cust_1641   3410.1575      0.10   
8   Ord_4725  Prod_13  SHP_6593  Cust_1641    162.0000      0.01   
9   Ord_4725   Prod_6  SHP_6593  Cust_1641     57.2200      0.07   
10  Ord_4743   Prod_2  SHP_6615  Cust_1641   4072.0100      0.01   
11  Ord_1925   Prod_6  SHP_2637   Cust_708    465.9000      0.05   
12  Ord_2978  Prod_16  SHP_4112  Cust_1088    305.0500      0.04   
13  Ord_2207  Prod_11  SHP_3093   Cust_839   3364.2480      0.10   
14  Ord_2207  Prod_10  SHP_3006   Cust_839   1410.9300      0.08   
15  Ord_2280   Prod_5  SHP_3114   Cust_839    460.6900      0.06   
16  Ord_2282   Prod_9  SHP_3122   Cust_839    443.4600      0.06   
17  Ord_4471  Prod_15  SHP_6228  Cust_1521  13255.9300      0.02   
18  Ord_4427   Prod_6  SHP_6171  Cust_1521    283.1300      0.08   
19   Ord_996  Prod_13  SHP_1378   Cust_371     41.9700      0.05   

    Order_Quantity   Profit  Shipping_Cost  Product_Base_Margin  
0               23   -30.51           3.60                 0.56  
1               13     4.56           0.93                 0.54  
2               26  1148.90           2.50                 0.59  
3               43   729.34          14.30                 0.37  
4               35  1219.87          26.30                 0.38  
5               23   -47.64           6.15                 0.37  
6                5     1.32           0.50                 0.36  
7               48  1137.91           0.99                 0.55  
8               33    45.84           0.71                 0.52  
9                8   -27.72           6.60                 0.37  
10              43  1675.98           0.99                 0.56  
11              38    79.34           4.86                 0.38  
12              27    23.12           3.37                 0.57  
13              15  -693.23          61.76                 0.78  
14              10  -317.48          36.09                 0.77  
15              48  -103.48           7.29                 0.45  
16              30   193.12           1.39                 0.38  
17              25  4089.27          26.00                 0.60  
18              45  -141.26           6.81                 0.36  
19              12   -37.03           4.20                 0.56

Now, arbitrary numeric indices are difficult to read and work with. Thus, you may want to change the indices of the df to something more meanigful.

Let's change the index to Ord_id (unique id of each order), so that you can select rows using the order ids directly.

In [ ]:
market_df.set_index('Ord_id', inplace=True)
market_df.head()

Having meaningful row labels as indices helps you to select (subset) dataframes easily. We will study selecting dataframes in the next section.


#### Sorting dataframes
You can sort dataframes in two ways 
- 1: by the indices and 
- 2: by the values.

In [30]:
market_df.head()

Prod_id   Ship_id    Cust_id    Sales  Discount  Order_Quantity  \
Ord_id                                                                      
Ord_5446  Prod_16  SHP_7609  Cust_1818   136.81      0.01              23   
Ord_5406  Prod_13  SHP_7549  Cust_1818    42.27      0.01              13   
Ord_5446   Prod_4  SHP_7610  Cust_1818  4701.69      0.00              26   
Ord_5456   Prod_6  SHP_7625  Cust_1818  2337.89      0.09              43   
Ord_5485  Prod_17  SHP_7664  Cust_1818  4233.15      0.08              35   

           Profit  Shipping_Cost  Product_Base_Margin  
Ord_id                                                 
Ord_5446   -30.51           3.60                 0.56  
Ord_5406     4.56           0.93                 0.54  
Ord_5446  1148.90           2.50                 0.59  
Ord_5456   729.34          14.30                 0.37  
Ord_5485  1219.87          26.30                 0.38

In [32]:
# sorting by index
# axis = 0 indicates to sort rows(use axis=1 for columns)
market_df.sort_index(axis=0, ascending = False)

Prod_id   Ship_id   Cust_id     Sales  Discount  Order_Quantity  \
Ord_id                                                                      
Ord_999   Prod_15  SHP_1383  Cust_361   5661.08      0.00              33   
Ord_998    Prod_8  SHP_1380  Cust_372    750.66      0.00              33   
Ord_998    Prod_5  SHP_1382  Cust_372   2149.37      0.03              42   
Ord_998    Prod_8  SHP_1381  Cust_372    254.32      0.01               8   
Ord_997   Prod_14  SHP_1379  Cust_365  28761.52      0.04               8   
...           ...       ...       ...       ...       ...             ...   
Ord_1001   Prod_5  SHP_1385  Cust_374   1981.26      0.07              49   
Ord_1000   Prod_6  SHP_1384  Cust_373    334.71      0.01              25   
Ord_100    Prod_8   SHP_138   Cust_58    121.12      0.10               3   
Ord_10     Prod_3    SHP_13   Cust_10     80.61      0.02              15   
Ord_1      Prod_1     SHP_1    Cust_1    261.54      0.04               6   

           Profit  Shipping_Cost  Product_Base_Margin  
Ord_id                                                 
Ord_999   1055.47          30.00                 0.62  
Ord_998    120.05           4.00                 0.60  
Ord_998    217.87          19.99                 0.55  
Ord_998   -117.39           6.50                 0.79  
Ord_997    285.11          24.49                 0.37  
...           ...            ...                  ...  
Ord_1001   100.80           8.66                 0.76  
Ord_1000    31.74           6.47                 0.38  
Ord_100   -118.82           1.99                 0.44  
Ord_10      -4.72           2.99                 0.37  
Ord_1     -213.25          35.00                 0.80  

[8399 rows x 9 columns]

In [35]:
# sorting by values
# sorting in increasing order of sales
market_df.sort_values(by='Sales').head()

Prod_id   Ship_id    Cust_id  Sales  Discount  Order_Quantity  \
Ord_id                                                                    
Ord_704    Prod_7   SHP_964   Cust_242   2.24      0.01               1   
Ord_149    Prod_3  SHP_7028  Cust_1712   3.20      0.09               1   
Ord_4270   Prod_7  SHP_5959  Cust_1450   3.23      0.06               2   
Ord_4755  Prod_13  SHP_6628  Cust_1579   3.41      0.06               1   
Ord_2252   Prod_3  SHP_3064   Cust_881   3.42      0.05               1   

          Profit  Shipping_Cost  Product_Base_Margin  
Ord_id                                                
Ord_704    -1.97           0.70                 0.37  
Ord_149    -3.16           1.49                 0.37  
Ord_4270   -2.73           0.70                 0.81  
Ord_4755   -1.78           0.70                 0.56  
Ord_2252   -2.91           1.49                 0.37

In [37]:
# sorting in decreasing order of Shipping_cost
market_df.sort_values(by='Shipping_Cost', ascending = False)

Prod_id   Ship_id    Cust_id      Sales  Discount  Order_Quantity  \
Ord_id                                                                        
Ord_1751  Prod_15  SHP_2426   Cust_597  14740.510      0.00              46   
Ord_839   Prod_11  SHP_1361   Cust_364  12689.870      0.04              44   
Ord_1741  Prod_11  SHP_2411   Cust_595  15168.820      0.02              26   
Ord_417   Prod_11   SHP_561   Cust_156  20333.816      0.02              45   
Ord_1581  Prod_15  SHP_2184   Cust_519   2573.920      0.07              17   
...           ...       ...        ...        ...       ...             ...   
Ord_4822  Prod_12  SHP_6718  Cust_1654     21.550      0.09               3   
Ord_2375  Prod_12  SHP_3263   Cust_939     90.150      0.07              13   
Ord_3324  Prod_12  SHP_4612  Cust_1178    154.240      0.03              31   
Ord_284   Prod_12   SHP_384   Cust_105     97.130      0.05               1   
Ord_3499  Prod_12  SHP_4846  Cust_1223     86.290      0.02              17   

           Profit  Shipping_Cost  Product_Base_Margin  
Ord_id                                                 
Ord_1751  3407.73         164.73                 0.56  
Ord_839   -169.23         154.12                 0.76  
Ord_1741 -1096.78         147.12                 0.80  
Ord_417  -1430.45         147.12                 0.80  
Ord_1581   117.23         143.71                 0.55  
...           ...            ...                  ...  
Ord_4822    -2.87           0.49                 0.38  
Ord_2375    33.05           0.49                 0.38  
Ord_3324    68.88           0.49                 0.39  
Ord_284    -37.06           0.49                 0.36  
Ord_3499    34.98           0.49                 0.39  

[8399 rows x 9 columns]

In [38]:
market_df.sort_values(by='Shipping_Cost', ascending = False).head()

Prod_id   Ship_id   Cust_id      Sales  Discount  Order_Quantity  \
Ord_id                                                                       
Ord_1751  Prod_15  SHP_2426  Cust_597  14740.510      0.00              46   
Ord_839   Prod_11  SHP_1361  Cust_364  12689.870      0.04              44   
Ord_1741  Prod_11  SHP_2411  Cust_595  15168.820      0.02              26   
Ord_417   Prod_11   SHP_561  Cust_156  20333.816      0.02              45   
Ord_1581  Prod_15  SHP_2184  Cust_519   2573.920      0.07              17   

           Profit  Shipping_Cost  Product_Base_Margin  
Ord_id                                                 
Ord_1751  3407.73         164.73                 0.56  
Ord_839   -169.23         154.12                 0.76  
Ord_1741 -1096.78         147.12                 0.80  
Ord_417  -1430.45         147.12                 0.80  
Ord_1581   117.23         143.71                 0.55

In [40]:
# sorting by more than two columns

# sorting in ascending order of sales for each product
market_df.sort_values(by=['Prod_id','Sales'],ascending=False)

Prod_id   Ship_id    Cust_id    Sales  Discount  Order_Quantity  \
Ord_id                                                                     
Ord_2197  Prod_9  SHP_2994   Cust_827  7522.80      0.04              48   
Ord_4356  Prod_9  SHP_6074  Cust_1481  6831.72      0.01              41   
Ord_262   Prod_9   SHP_358    Cust_66  6553.45      0.03              39   
Ord_4059  Prod_9  SHP_5660  Cust_1378  5587.20      0.05              36   
Ord_2973  Prod_9  SHP_6073  Cust_1480  5410.95      0.09              36   
...          ...       ...        ...      ...       ...             ...   
Ord_3713  Prod_1  SHP_5145  Cust_1307    27.83      0.09               2   
Ord_2746  Prod_1  SHP_3767  Cust_1030    22.61      0.03               1   
Ord_439   Prod_1   SHP_587   Cust_136    18.73      0.05               1   
Ord_2314  Prod_1  SHP_3171   Cust_899    18.16      0.03               1   
Ord_286   Prod_1   SHP_387    Cust_90    18.15      0.04               1   

           Profit  Shipping_Cost  Product_Base_Margin  
Ord_id                                                 
Ord_2197  3187.37          19.99                 0.39  
Ord_4356  3081.02          19.99                 0.39  
Ord_262   2969.81          19.99                 0.39  
Ord_4059  2254.16          19.99                 0.39  
Ord_2973  2077.91          19.99                 0.39  
...           ...            ...                  ...  
Ord_3713   -22.14           9.45                 0.60  
Ord_2746    -8.40           7.51                 0.57  
Ord_439     -6.68           6.13                 0.57  
Ord_2314    -7.25           6.13                 0.57  
Ord_286     -7.26           6.13                 0.57  

[8399 rows x 9 columns]

In [41]:
# sorting in ascending order of sales for each product
market_df.sort_values(by=['Prod_id','Sales'],ascending=False).head()


Prod_id   Ship_id    Cust_id    Sales  Discount  Order_Quantity  \
Ord_id                                                                     
Ord_2197  Prod_9  SHP_2994   Cust_827  7522.80      0.04              48   
Ord_4356  Prod_9  SHP_6074  Cust_1481  6831.72      0.01              41   
Ord_262   Prod_9   SHP_358    Cust_66  6553.45      0.03              39   
Ord_4059  Prod_9  SHP_5660  Cust_1378  5587.20      0.05              36   
Ord_2973  Prod_9  SHP_6073  Cust_1480  5410.95      0.09              36   

           Profit  Shipping_Cost  Product_Base_Margin  
Ord_id                                                 
Ord_2197  3187.37          19.99                 0.39  
Ord_4356  3081.02          19.99                 0.39  
Ord_262   2969.81          19.99                 0.39  
Ord_4059  2254.16          19.99                 0.39  
Ord_2973  2077.91          19.99                 0.39

In [23]:
# sorting in ascending order of sales for each product
market_df.sort_values(by=['Prod_id','Sales'],ascending=False).tail()

Prod_id   Ship_id    Cust_id  Sales  Discount  Order_Quantity  \
Ord_id                                                                   
Ord_3713  Prod_1  SHP_5145  Cust_1307  27.83      0.09               2   
Ord_2746  Prod_1  SHP_3767  Cust_1030  22.61      0.03               1   
Ord_439   Prod_1   SHP_587   Cust_136  18.73      0.05               1   
Ord_2314  Prod_1  SHP_3171   Cust_899  18.16      0.03               1   
Ord_286   Prod_1   SHP_387    Cust_90  18.15      0.04               1   

          Profit  Shipping_Cost  Product_Base_Margin  
Ord_id                                                
Ord_3713  -22.14           9.45                 0.60  
Ord_2746   -8.40           7.51                 0.57  
Ord_439    -6.68           6.13                 0.57  
Ord_2314   -7.25           6.13                 0.57  
Ord_286    -7.26           6.13                 0.57